### PR TITLE
[GS-124] Automatically generate gradebook AND CSV-ify PL grades for one assignment and push them to a sheet

### PR DESCRIPTION
**Pull request by naveen-nathan**
### Jira Ticket

[Jira Ticket](https://dashbd.atlassian.net/browse/GS-124)

### Description

An automatically generated gradebook

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would change existing functionality)

### Changes

Added gradebook sheets which retrieve data for assignments from the corresponding subsheets.

### Testing

Ran this locally.

Here is a [test sheet](https://docs.google.com/spreadsheets/d/1daoJb0W_ljrVIvtKOhG4jf8aE3uVUHTJ2gbrexcV3Ps/edit?gid=0#gid=0) generated from scratch by this version of script.

### Checklist

- [X] My branch name matches the format: `<ticket-id>/<brief-description-of-change>`
- [X] My PR name matches the format: `[<ticket-id>] <brief-description-of-change>`
- [ ] I have added doc-comments to all new functions ([JSDoc](https://jsdoc.app/) for JS and [Docstrings](https://peps.python.org/pep-0257/) for Python)
- [ ] I have reviewed all of my code
- [X] My code only contains major changes related to my ticket

### CODE REVIEW

File: gradescopeCronJob/README.md
Comment (Mehul-Gandhi) [inline gradescopeCronJob/README.md]:
```suggestion
# 5. Docker Container
```

File: gradescopeCronJob/README.md
Comment (Mehul-Gandhi) [inline gradescopeCronJob/README.md:75]: It might make more sense for this to go before step 3, defining a config file, because we need the spreadsheet ID before we can input the ID into the config file

File: gradescopeCronJob/README.md
Comment (Mehul-Gandhi) [inline gradescopeCronJob/README.md]: Where does a TA/instructor obtain this roster in this format? Is this easy to do?

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [inline gradescopeCronJob/gradescope_to_spreadsheet.py]: Can you make a short comment explaining this?

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [inline gradescopeCronJob/gradescope_to_spreadsheet.py:296]: Can we rename `initialize_gs_client` to `initialize_gradescope_client`? This is because GS could mean: gradesync, Google Sheets, or Gradescope

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [inline gradescopeCronJob/gradescope_to_spreadsheet.py]:
```suggestion
    get_sub_sheet_titles_to_ids(sheet_api_instance)
```

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [inline gradescopeCronJob/gradescope_to_spreadsheet.py:303]:
```suggestion
    make_batch_request(sheet_api_instance)
```

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [inline gradescopeCronJob/gradescope_to_spreadsheet.py]:
```suggestion
@deprecated
```

Comment (Mehul-Gandhi) [reply gradescopeCronJob/gradescope_to_spreadsheet.py]: 1. Pandas is already imported, so we don't need this reminder.
2. We can use the @deprecated decorator here

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [inline gradescopeCronJob/gradescope_to_spreadsheet.py:443]: Can you revise this docstring to include information about the instructor dashboard?

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [inline gradescopeCronJob/gradescope_to_spreadsheet.py:390]: Can you make a docstring for this function?

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [inline gradescopeCronJob/gradescope_to_spreadsheet.py:365]: Can you make a docstring and also explain the logic that is happening with `formula_list`?

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [reply gradescopeCronJob/gradescope_to_spreadsheet.py:443]: Can you explain the new flow? From my interpretation, the new flow is :
1. Create subsheets for each assignment in a single batch request (this was the old approach and is still included in this new approach)
2. Create a subsheet for the "instructor dashboard", which has all information on a single subsheet. There are 4 total API calls to google spreadsheet, one for each category type. This subsheet depends on step 1 being processed successfully.

Can you document or list the total amount of API calls we are making to Google Sheets?

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [inline gradescopeCronJob/gradescope_to_spreadsheet.py:372]: suggestion: include a logging statement such as f"Successfully processed {category} on the instructor dashboard". The logging message in `assemble_rest_request_for_assignment` is not detailed enough to help with debugging since there are 4 different categories we are processing.

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [inline gradescopeCronJob/gradescope_to_spreadsheet.py:291]: In the function `retrieve_grades_from_gradescope`, can you add a logging statement that will log if retrieving the grades from gradescope succeeded or failed?

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [inline gradescopeCronJob/gradescope_to_spreadsheet.py]: Can you make a docstring for this function?

File: gradescopeCronJob/gradescope_to_spreadsheet.py
Comment (Mehul-Gandhi) [inline gradescopeCronJob/gradescope_to_spreadsheet.py:87]: Remind us through comments on what the criteria is for getting discussion credit, or what this formula does.

**Comment by naveen-nathan** [comment]
Just added functionality for a discussion gradebook tab; I updated the test sheet accordingly.

**Review by Mehul-Gandhi — DISMISSED** [review]
Requirements:
1. Add docstrings. The most important part is explaining line 270 and the function on line 272. And line 70-71.
2. Comment on the runtime of this script when you tested and the number of API requests being made to Google Sheets
___
Comment: The result looks good, but to be honest I am not sure how to see if the instructor dashboard is referencing the right values.
___
Questions:
1. What is row 193 trying to do in the "Midterms"/"Labs" tab?
2. Can you give an example of how to interpret something like 
`=DIVIDE(XLOOKUP(C:C, INDIRECT( INDIRECT(ADDRESS(1, COLUMN(), 4)) & "!C:C"), INDIRECT(INDIRECT(ADDRESS(1, COLUMN(), 4)) & "!F:F")), XLOOKUP(C:C, INDIRECT( INDIRECT(ADDRESS(1, COLUMN(), 4)) & "!C:C"), INDIRECT(INDIRECT(ADDRESS(1, COLUMN(), 4)) & "!G:G")))`?
This can be done in PR comments.
___
 Comment: I will let you check in with Dan to see if that is the format he wanted. I can approve the PR after the docstrings are added, and you can merge it before or after Dan approves the Sheets format.

**Comment by Mehul-Gandhi** [comment]
If you are still adding changes to this PR, then set the PR to draft mode and then request a review when you are ready. Or make another branch/ticket to continue your work on adding more features.

File: gradescopeCronJob/README.md
Comment (naveen-nathan) [reply gradescopeCronJob/README.md:75]: Very good point! Will change the README accordingly.

File: gradescopeCronJob/README.md
Comment (naveen-nathan) [reply gradescopeCronJob/README.md]: They can get it from bCourses. Good question; will specify this in the README. It is easy to do.

**Review by Mehul-Gandhi — APPROVED** [review]
Emergency merge

**Pull request merged**
